src/classes/rest_server.vala

```diff
@@ -0,0 +1,594 @@
+                var body = msg.request_body;
+                Json.Parser parser = new Json.Parser();
+                try {
+                    parser.load_from_data((string) body.data,
+                        (ssize_t) body.length);
+                    Json.Node node = parser.get_root();
+                    if (node != null && node.get_node_type() == Json.NodeType.OBJECT) {
+                        unowned Json.Object obj = node.get_object();
+                        string action = null, argument = null;
+                        foreach (unowned string name in obj.get_members()) {
+                            unowned Json.Node item = obj.get_member(name);
+                            switch (name) {
+                            case "action":
+                                action = item.get_string();
+                                break;
+                            case "argument":
+                                argument = item.get_string();
+                                break;
+                            }
+                        }
+                
+                        if (action != null) {
+                            if (argument == "" || argument == null) {
+                                this.metadata.controller.trigger_action(action,
+                                    null);
+                            } else {
+                                this.metadata.controller.trigger_action(action,
+                                    argument);
+                            }
+                            // FIXME: use a sound logic or separate REST path
+                            if (action == "movePointer") {
+                                root = this.pointer_data();
+                            } else {
+                                root = this.state_data();
+                            }
+                        } else {
+                            root = this.error_data("Action not defined");
+                            msg.status_code = 400;
+                        }
+                    } else {
+                        root = this.error_data("Invalid request");
+                        msg.status_code = 400;
+                    }
+                } catch (Error e) {
+                    root = this.error_data(e.message);
+                    msg.status_code = 400;
+                }
+            } else {
+                root = this.error_data("Not found");
+                msg.status_code = 404;
+            }
+
+            if (root != null) {
+                Json.Generator generator = new Json.Generator();
+                generator.set_root(root);
+                string response = generator.to_data(null);
+
+                msg.set_response("application/json", Soup.MemoryUse.COPY,
+                    response.data);
+            }
+        }
+
+        public RestServer(Metadata.Pdf metadata, int port_num) {
+            TlsCertificate cert = null;
+            if (Options.rest_https) {
```

Comment:
If I understand it correctly, you would use the same port for HTTP and HTTPS traffic. Is this intentional?

Reply:
Yes. It is anyway not possible to run HTTP and HTTPS simultaneously (can be implemented, but I see absolutely no need), so why make it more complicated?